File: PA41.java

//Leetcode 105. Construct Binary Tree from Preorder and Inorder Traversal
//Recursive solution
Time complexity : O(n) // where n is total number of nodes in tree 
Space Complexity: O(2n) // n for Hashset and n for recursive stack
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int index=0;
    HashMap<Integer, Integer> hm;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if(preorder==null || preorder.length==0 || inorder==null || inorder.length==0  ){
            return null;
        }
        
        hm = new HashMap<>();
        
        for(int i=0;i<inorder.length ;i++){
            hm.put(inorder[i],i);
            
        }
        return helper(preorder, inorder, 0, inorder.length-1);
        
    }
    public TreeNode helper(int[] preoder,int[] inorder, int start, int end){
        if(start>end || index>preoder.length-1){
            return null;
        
        }
        int rval= preoder[index];
        TreeNode root= new TreeNode(rval);
        index++;
        int m= hm.get(rval);
        
        if(start== end){
            return root;
        }
        root.left= helper(preoder, inorder,start, m-1);
        root.right= helper(preoder, inorder,m+1, end);
        
        return root;
        
        
        
    }
}

File: PA_41.py
#LC 105 - Construct Binary Tree from Preorder and Inorder Traversal
#Time Complexity - O(n)
#Space Complexity - O(n)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        
        self.d = {}
        self.index = 0
        for i in range(len(inorder)):
            self.d[inorder[i]] = i
        def helper(inorder,preorder,start,end):
            if self.index >= len(preorder) or start > end:
                return None
        
            root = TreeNode(preorder[self.index])
            self.index = self.index + 1
            inIdx = self.d[root.val]
            root.left = helper(inorder,preorder,start,inIdx-1)
            root.right = helper(inorder,preorder,inIdx+1,end)
            return root
        return helper(inorder,preorder,0,len(inorder)-1)